### ISSUE TITLE
Make 'return' from a quasi DTRT

### ISSUE THREAD

**Issue by masak** [masak/007#240]
Following up on discussion that fell out from #233 (now fixed):

    <masak> I think we might want to open up a separate issue for `return` in `quasi`, though:
    <masak> $ bin/007 -e='macro moo() { quasi { say(1); return; say(2) } }; moo()'
    <masak> 1
    <masak> Variable '--RETURN-TO--' is not declared
    <masak> [...]
    <masak> (wow, that's a really crappy error message, even)

I'd like us, in the short term, to have a better runtime error here, same as the parse-time one we have:

    $ bin/007 -e='return'
    Attempt to return outside of any Routine

But in the long term, I'd like for us to be able to "return" from quasis. This might in fact become a lot easier once `Q::Expr::BlockAdapter` becomes a thing.

**Comment by vendethiel** [comment]
> But in the long term, I'd like for us to be able to "return" from quasis.

I think we discussed this already, but I feel very strongly about it: a `return` in a quasi means you get a `return` once "instantiated".

**Comment by masak** [comment]
@vendethiel I think I understand what you mean, but could you please show how you prefer it to work with a minimal piece of code?

Does this cover it?

    macro moo() {
        quasi {
            return 42;
        }
    }

    function foo() {
        moo();
        return 9;
    }

    say(foo());    # current semantics: error; masak suggests: 9; vendethiel suggests: 42

Did I get that right?

**Comment by vendethiel** [comment]
Yes, that's correct.

**Comment by masak** [comment]
Then I'll object in the same way I probably did that other time: `return` in Perl 6 is *lexical*; that is, we expect the same kind of hygiene from it as we expect from variables. 007 is trying to emulate Perl 6 here. The "42" behavior is essentially unhygienic. The "9" behavior is inconsistent but "strangely consistent".

I want to add that I understand why you're arguing for the "42" behavior. We should be looking at having some way to cater to that use case. But I don't think it should be a plain `return` &mdash; maybe it should be a combination of `return` and a ([S06](http://design.perl6.org/S06.html#Splicing)) `COMPILING::` declaration. Would that feel OK?

**Comment by vendethiel** [comment]
It does not. If anything, then a stray return in a quasi should be an error. "Quasi code does not run."

**Comment by masak** [comment]
Yes, that's essentially the current semantics.

I'm keen on arriving at a mutually-satisfactory consensus with you on this one; perhaps the next time we meet on IRC we can hash it out together? Might be a little bit more wide-band than Github issue comments. Whatever we come up with, we can feed back into this issue.

**Comment by masak** [comment]
Hm. I remember us discussing this on IRC, but I don't recall us arriving at a good consensus.

@vendethiel, I'm not very inclined to go ahead implementing my desired semantics of `return` if it feels unexpected to you. On the other hand, your suggested semantics feels unexpected to me, and I feel I have good reasons for thinking it inconsistent.

So, for the time being, the status quo prevails. Closing issue.

**Comment by vendethiel** [comment]
I'm not sure we should close this issue anyway. Maybe we should at least provide a better error message in that case?

**Comment by masak** [comment]
Agreed. Opened #276 for the error message.

**Comment by masak** [comment]
Looking at this with fresh eyes some four years later. I would suggest this feature:

```
macro moo() {
    quasi {
        COMPILING.return 42;
    }
}

function foo() {
    moo();
    return 9;
}

say(foo());    # 42
```

That is, `COMPILING.return` is a "package-scoped keyword" expressing that what you want to return from is the injectile's surrounding routine. It's a little bit long and ugly, but that's part way intentional (as "syntactic salt"), because the intent expressed is not on any kind of common path.

**Comment by vendethiel** [comment]
Re-reading this issue, I have a question…

> `return` in Perl 6 is *lexical*

Is that actually true? Isn’t it technically a control flow exception, and thus dynamic?

Related (but not exactly the same), how do you call the fact that simple closures return from their parent subs?

```perl
sub f(@x) { @x.map({ return }) }
```

**Comment by masak** [comment]
> Re-reading this issue, I have a question…
>
> > `return` in Perl 6 is *lexical*
>
> Is that actually true? Isn’t it technically a control flow exception, and thus dynamic?

Yes, you are right.

This actually came up at `$work`, in the context of what the following `return` in a parameter default ought to do:

```
sub foo($v = do return 1) {
    return 2;
}
```

(I wrote it as Raku code, so I added a `do` as an adapter from statements to expressions.)

Picture `return` being implemented as an effect, handled by an implicit effect handler set up by entering the function. (Or the corresponding terminology but with delimited continuations.) Does the above work &mdash; did we set up the handler already before entering the function body? (Remember, evaluating the arguments to bind to the parameters needs to happen in the lexical scope of the caller.)

What about if we make `foo` a `multi`, and throw in a number of other candidates?

I believe the answer here isn't "yes" or "no", but &mdash; as often happens in PLT &mdash; "I'm so confused now", followed by strong liquor if the situation permits. 

> Related (but not exactly the same), how do you call the fact that simple closures return from their parent subs?
>
> ```perl
> sub f(@x) { @x.map({ return }) }
> ```

In Perl 6/Raku, those are called "blocks" (as distinguished from "routines"); I've always thought of them as having slightly weaker "walls" so they don't have their own `return` handler. (They do have a `leave` handler, but I'm not sure that was ever fully implemented...? It certainly wasn't very frequently used.) In Alma, I've been trying to avoid (first-class) blocks, so that when you pass a routine to `map`, it's guaranteed to have its own `return`.
